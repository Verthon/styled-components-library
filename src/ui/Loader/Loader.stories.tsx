import { ComponentStory, ComponentMeta } from '@storybook/react'

import Loader from './Loader'

export default {
  title: 'ui/Loader',
  component: Loader,
} as ComponentMeta<typeof Loader>

const Template: ComponentStory<typeof Loader> = (args) => <Loader {...args} />

export const Default = Template.bind({})

export const CustomDimensions = Template.bind({})
CustomDimensions.args = {
  width: 100,
  height: 100,
}

export const CustomThickness = Template.bind({})
CustomThickness.args = {
  thickness: 5,
}
