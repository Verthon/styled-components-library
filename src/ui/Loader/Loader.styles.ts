import styled, { css } from 'styled-components'

import type { InnerLoaderStyleProps, LoaderStyleProps } from './Loader.types'

const Wrapper = styled.div<LoaderStyleProps>(
  ({ width, height, thickness, color }) => css`
    @keyframes spinner {
      0% {
        transform: rotate(0deg);
      }
      100% {
        transform: rotate(360deg);
      }
    }
    position: relative;
    height: ${width}px;
    width: ${height}px;
    margin: auto;

    &:before {
      content: '';
      position: absolute;
      animation-delay: -0.15s;
      width: 80%;
      height: 80%;
      border: ${thickness}px solid transparent;
      border-top-color: ${color};
      border-radius: 50%;
      animation: spinner 1.2s cubic-bezier(0.5, 0, 0.5, 1) infinite;
      top: 10%;
      left: 10%;
    }

    &:after {
      content: '';
      position: absolute;
      width: 80%;
      height: 80%;
      border: ${thickness}px solid transparent;
      border-top-color: ${color};
      border-radius: 50%;
      animation: spinner 1.2s cubic-bezier(0.5, 0, 0.5, 1) infinite;
      top: 10%;
      left: 10%;
    }
  `,
)

Wrapper.displayName = 'LLoader--Wrapper'

const InnerElement = styled.span<InnerLoaderStyleProps>(
  ({ thickness, color }) => css`
    position: absolute;
    top: 0;
    bottom: 0;
    right: 0;
    left: 0;

    &:before {
      content: '';
      position: absolute;
      width: 80%;
      height: 80%;
      border: ${thickness}px solid transparent;
      border-top-color: ${color};
      border-radius: 50%;
      animation-delay: -0.45s;
      animation: spinner 1.2s cubic-bezier(0.5, 0, 0.5, 1) infinite;
      top: 10%;
      left: 10%;
    }

    &:after {
      content: '';
      position: absolute;
      width: 80%;
      height: 80%;
      border: ${thickness}px solid transparent;
      border-top-color: ${color};
      border-radius: 50%;
      animation-delay: -0.3s;
      animation: spinner 1.2s cubic-bezier(0.5, 0, 0.5, 1) infinite;
      top: 10%;
      left: 10%;
    }
  `,
)

InnerElement.displayName = 'LLoader--InnerElement'

export const Styled = {
  Wrapper,
  InnerElement,
}
