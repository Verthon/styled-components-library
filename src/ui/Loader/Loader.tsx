import { Styled } from './Loader.styles'
import type { LoaderProps } from './Loader.types'

const DEFAULT_HEIGHT = 200
const DEFAULT_WIDTH = 200
const THICKNESS = {
  sm: 8,
  md: 14,
  lg: 20,
}

const Loader = ({ width, height, thickness, className, color }: LoaderProps) => {
  const currentThickness = thickness || THICKNESS['md']

  return (
    <Styled.Wrapper
      width={width || DEFAULT_WIDTH}
      height={height || DEFAULT_HEIGHT}
      thickness={currentThickness}
      className={className}
      role="progressbar"
      color={color || 'inherit'}
    >
      <Styled.InnerElement thickness={currentThickness} color={color || 'inherit'} />
    </Styled.Wrapper>
  )
}

export default Loader
