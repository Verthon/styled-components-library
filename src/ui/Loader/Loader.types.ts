export type LoaderProps = {
  width?: number
  height?: number
  className?: string
  thickness?: number
  color?: string
}

export type LoaderStyleProps = Required<Pick<LoaderProps, 'height' | 'width' | 'thickness' | 'color'>>

export type InnerLoaderStyleProps = Pick<LoaderStyleProps, 'thickness' | 'color'>
