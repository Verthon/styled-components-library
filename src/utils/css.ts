import type { ValueInRems } from '../types/css'

export const getValueFromRem = (remValue: ValueInRems) => {
  const number = remValue.replace('rem', '')
  return Number(number)
}
