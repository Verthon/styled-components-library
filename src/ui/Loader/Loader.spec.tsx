import { screen } from '@testing-library/dom'

import { createTestWrapper } from '../../tests/utils'

import Loader from './Loader'

describe('Loader', () => {
  it('should render Loader', () => {
    createTestWrapper({ children: <Loader /> })

    expect(screen.getByRole('progressbar')).toBeInTheDocument()
  })

  it('should render Loader with different height and width and thickness', () => {
    const loader = createTestWrapper({ children: <Loader width={100} height={100} thickness={4} /> })

    expect(loader).toMatchSnapshot()
  })
})
