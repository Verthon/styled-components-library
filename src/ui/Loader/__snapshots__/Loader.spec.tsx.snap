// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader should render Loader with different height and width and thickness 1`] = `
{
  "asFragment": [Function],
  "baseElement": .c0 {
  position: relative;
  height: 100px;
  width: 100px;
  margin: auto;
}

.c0:before {
  content: '';
  position: absolute;
  -webkit-animation-delay: -0.15s;
  animation-delay: -0.15s;
  width: 80%;
  height: 80%;
  border: 4px solid transparent;
  border-top-color: inherit;
  border-radius: 50%;
  -webkit-animation: spinner 1.2s cubic-bezier(0.5,0,0.5,1) infinite;
  animation: spinner 1.2s cubic-bezier(0.5,0,0.5,1) infinite;
  top: 10%;
  left: 10%;
}

.c0:after {
  content: '';
  position: absolute;
  width: 80%;
  height: 80%;
  border: 4px solid transparent;
  border-top-color: inherit;
  border-radius: 50%;
  -webkit-animation: spinner 1.2s cubic-bezier(0.5,0,0.5,1) infinite;
  animation: spinner 1.2s cubic-bezier(0.5,0,0.5,1) infinite;
  top: 10%;
  left: 10%;
}

.c1 {
  position: absolute;
  top: 0;
  bottom: 0;
  right: 0;
  left: 0;
}

.c1:before {
  content: '';
  position: absolute;
  width: 80%;
  height: 80%;
  border: 4px solid transparent;
  border-top-color: inherit;
  border-radius: 50%;
  -webkit-animation-delay: -0.45s;
  animation-delay: -0.45s;
  -webkit-animation: spinner 1.2s cubic-bezier(0.5,0,0.5,1) infinite;
  animation: spinner 1.2s cubic-bezier(0.5,0,0.5,1) infinite;
  top: 10%;
  left: 10%;
}

.c1:after {
  content: '';
  position: absolute;
  width: 80%;
  height: 80%;
  border: 4px solid transparent;
  border-top-color: inherit;
  border-radius: 50%;
  -webkit-animation-delay: -0.3s;
  animation-delay: -0.3s;
  -webkit-animation: spinner 1.2s cubic-bezier(0.5,0,0.5,1) infinite;
  animation: spinner 1.2s cubic-bezier(0.5,0,0.5,1) infinite;
  top: 10%;
  left: 10%;
}

<body>
    <div>
      <div
        class="c0"
        color="inherit"
        height="100"
        role="progressbar"
        width="100"
      >
        <span
          class="c1"
          color="inherit"
        />
      </div>
    </div>
  </body>,
  "container": .c0 {
  position: relative;
  height: 100px;
  width: 100px;
  margin: auto;
}

.c0:before {
  content: '';
  position: absolute;
  -webkit-animation-delay: -0.15s;
  animation-delay: -0.15s;
  width: 80%;
  height: 80%;
  border: 4px solid transparent;
  border-top-color: inherit;
  border-radius: 50%;
  -webkit-animation: spinner 1.2s cubic-bezier(0.5,0,0.5,1) infinite;
  animation: spinner 1.2s cubic-bezier(0.5,0,0.5,1) infinite;
  top: 10%;
  left: 10%;
}

.c0:after {
  content: '';
  position: absolute;
  width: 80%;
  height: 80%;
  border: 4px solid transparent;
  border-top-color: inherit;
  border-radius: 50%;
  -webkit-animation: spinner 1.2s cubic-bezier(0.5,0,0.5,1) infinite;
  animation: spinner 1.2s cubic-bezier(0.5,0,0.5,1) infinite;
  top: 10%;
  left: 10%;
}

.c1 {
  position: absolute;
  top: 0;
  bottom: 0;
  right: 0;
  left: 0;
}

.c1:before {
  content: '';
  position: absolute;
  width: 80%;
  height: 80%;
  border: 4px solid transparent;
  border-top-color: inherit;
  border-radius: 50%;
  -webkit-animation-delay: -0.45s;
  animation-delay: -0.45s;
  -webkit-animation: spinner 1.2s cubic-bezier(0.5,0,0.5,1) infinite;
  animation: spinner 1.2s cubic-bezier(0.5,0,0.5,1) infinite;
  top: 10%;
  left: 10%;
}

.c1:after {
  content: '';
  position: absolute;
  width: 80%;
  height: 80%;
  border: 4px solid transparent;
  border-top-color: inherit;
  border-radius: 50%;
  -webkit-animation-delay: -0.3s;
  animation-delay: -0.3s;
  -webkit-animation: spinner 1.2s cubic-bezier(0.5,0,0.5,1) infinite;
  animation: spinner 1.2s cubic-bezier(0.5,0,0.5,1) infinite;
  top: 10%;
  left: 10%;
}

<div>
    <div
      class="c0"
      color="inherit"
      height="100"
      role="progressbar"
      width="100"
    >
      <span
        class="c1"
        color="inherit"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
  "user": {
    "clear": [Function],
    "click": [Function],
    "copy": [Function],
    "cut": [Function],
    "dblClick": [Function],
    "deselectOptions": [Function],
    "hover": [Function],
    "keyboard": [Function],
    "paste": [Function],
    "pointer": [Function],
    "selectOptions": [Function],
    "setup": [Function],
    "tab": [Function],
    "tripleClick": [Function],
    "type": [Function],
    "unhover": [Function],
    "upload": [Function],
  },
}
`;
